### PR TITLE
test(Cilium): add CiliumAgentNotReadyTaint to prevent overprovisioning

### PR DESCRIPTION
**Pull request by charliedmcb**
Fixes # 

**Description**
See the following thread for why we need the given startupTaint:
- https://karpenter.sh/docs/faq/#why-do-i-see-extra-nodes-get-launched-to-schedule-pending-pods-that-remain-empty-and-are-later-removed

As our testing, and main example cases all use Cilium, adding this to all the areas.

**How was this change tested?**
Tested [general-purpose.yaml](https://github.com/Azure/karpenter-provider-azure/pull/508/files#diff-d76365775af789aed6414b86f28136204ea77b35ee25349d76eb590b9d7c03b9) in self-hosted codespace setup.

Also, ran E2E suite:
- https://github.com/Azure/karpenter-provider-azure/actions/runs/11149258534
- https://github.com/Azure/karpenter-provider-azure/actions/runs/11150952645

**Does this change impact docs?**
- [ ] Yes, PR includes docs updates
- [ ] Yes, issue opened: # 
- [x] No

**Release Note**

```release-note
NONE
```

### CODE REVIEW

**Comment by tallaxes** [comment]
If this is always required when running with Cilium dataplane, can we automatically (conditional on Cilium dataplane) add this taint to the taints we pass to kubelet internally, rather than requiring explicit external configuration?

**Comment by charliedmcb** [comment]
> If this is always required when running with Cilium dataplane, can we automatically (conditional on Cilium dataplane) add this taint to the taints we pass to kubelet internally, rather than requiring explicit external configuration?

I don't think that would solve things, as the main issue here from my understanding is that the nodes come up with this taint already, but karpenter-core logic for provisioning isn't aware of it. Added it to the startuptaints, makes it so the taint will be there on startup, but also expected to be temporary, so won't mess up its provisioning logic.

See documentation:
- https://karpenter.sh/docs/concepts/nodepools/#cilium-startup-taint

**Comment by tallaxes** [comment]
Ok, so this is more about letting Karpenter know, than actually placing the taint. Would be nice to not have to configure this manually, but maybe it is unavoidable (at least in self-hosted).

**Review by tallaxes — APPROVED** [review]

File: examples/v1beta1/general-purpose.yaml [hunk 1]
```diff
@@ -11,6 +11,9 @@ spec:
     expireAfter: Never
   template:
     spec:
+      startupTaints:
+        - key: node.cilium.io/agent-not-ready
```
Comment (Bryce-Soghigian) [inline examples/v1beta1/general-purpose.yaml:15]:
```suggestion
        # If you are using cilium dataplane you might notice that there is some overprovisioning where we launch extra nodes to schedule pods that remain empty and are later removed. 
        # this taint will prevent that scenario, read the docs here: https://karpenter.sh/docs/faq/#why-do-i-see-extra-nodes-get-launched-to-schedule-pending-pods-that-remain-empty-and-are-later-removed
        - key: node.cilium.io/agent-not-ready
```

Comment (Bryce-Soghigian) [reply examples/v1beta1/general-purpose.yaml:15]: Maybe some context here to make it clear this is just if you are using cilium. network unavailable during cni initialization should also have this same side effect.

**Comment by charliedmcb** [comment]
> Ok, so this is more about letting Karpenter know, than actually placing the taint. Would be nice to not have to configure this manually, but maybe it is unavoidable (at least in self-hosted). And would need to be careful not to use this with non-cilium dataplanes.

Yea, I would like this to be something we can automatically configure at least within the e2e tests. However, isn't the simplest right now, and our general testing suggests going to cilium, so wanting to fix.

**Pull request merged**